<h1>Practice With Classification</h1>

In [1]:
import scipy
# TensorFFlow 2.10.0
import numpy as np

In [2]:
# Reference:  https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_classification.html#sklearn.datasets.make_classification


def generateData1():
    from sklearn.datasets import make_classification
    generated_data, generated_labels = make_classification(n_samples=40000, n_features=67,n_informative=6,
                                     n_redundant=21,n_classes=4,n_clusters_per_class=3,
                                     class_sep=0.85, random_state=4)
    return generated_data, generated_labels
#Generated a dataset for classification
generated_data, generated_labels=generateData1()

<h2>Printing the Data</h2>
<p>First going to print the data to be able to observe some of the features and entries. Then move forward to preprocessing.</p>

In [4]:
print(generated_data)

[[ 0.55886552 -1.0367371  -2.0716188  ... -1.66748974  0.88442319
   2.47414118]
 [-0.36649363  3.99879789 -0.94638273 ... -2.37814551 -0.76884534
  -1.0220301 ]
 [-1.0572081   0.97359163  0.34744335 ...  1.12169577 -2.40200316
   0.56310249]
 ...
 [ 0.78227173 -0.47473335 -1.01507181 ...  1.02149934  1.93028971
  -0.77386558]
 [ 0.65287134 -0.32653986 -0.5734769  ...  1.01438409 -1.29947182
  -1.83709253]
 [-0.33382408  0.8938808   0.02252798 ... -1.13487992  0.09518901
   0.58224631]]


In [5]:
import pandas as pd
from sklearn.preprocessing import StandardScaler
generated_data_dataframe = pd.DataFrame(data=generated_data)
scalerObject = StandardScaler().fit(generated_data_dataframe)
standardized_generated_dataframe = scalerObject.transform(generated_data_dataframe)
print(generated_data_dataframe)

             0         1         2         3         4         5         6   \
0      0.558866 -1.036737 -2.071619  2.615634  0.251788 -0.208035  2.895468   
1     -0.366494  3.998798 -0.946383 -0.635905  2.286920 -0.637968 -1.108372   
2     -1.057208  0.973592  0.347443 -1.655197  2.006350  0.794712 -0.900860   
3      1.386458 -0.462859  1.708425 -0.692245 -1.184841 -0.443195 -0.265234   
4     -0.675302 -1.535932  1.267888  1.639375 -0.236435 -1.173418  1.149061   
...         ...       ...       ...       ...       ...       ...       ...   
39995  0.853123 -3.544765  1.625326  0.128059  1.139331  0.728333  0.815937   
39996 -0.473624 -2.987561 -0.541470 -0.689567  0.253457 -0.221994  1.305095   
39997  0.782272 -0.474733 -1.015072  0.227328  0.263934  0.229079  2.961405   
39998  0.652871 -0.326540 -0.573477 -0.997861  2.130370 -0.381276 -2.629324   
39999 -0.333824  0.893881  0.022528  0.663176  1.381133  1.004063  0.839596   

             7         8         9   ...        57 

In [6]:
print(standardized_generated_dataframe)

[[ 0.56203181 -0.31366006 -2.0805219  ... -1.65797078  0.41370309
   2.47499216]
 [-0.3660114   1.65838212 -0.95197803 ... -2.36722943 -0.36008486
  -1.0264752 ]
 [-1.05872933  0.47363524  0.34565145 ...  1.12573131 -1.12446028
   0.56105859]
 ...
 [ 0.78608603 -0.09356526 -1.02086904 ...  1.02573186  0.90320548
  -0.77793476]
 [ 0.65631031 -0.03552896 -0.57797599 ...  1.0186306  -0.60843675
  -1.84277231]
 [-0.3332471   0.44241848  0.01978094 ... -1.12640805  0.04431367
   0.5802314 ]]


In [7]:
#https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.SelectKBest.html#sklearn.feature_selection.SelectKBest
from sklearn.feature_selection import SelectKBest
def featureSelectionOfKBestMeasure(submitted_data, submittedLabels, focused_number_of_features):
    for number in focused_number_of_features:
        transform_by_features = SelectKBest(k=number)
        transform_by_features.fit(submitted_data, submittedLabels)
        print(transform_by_features)
        transformed_data = transform_by_features.fit_transform(submitted_data, submittedLabels)
        print("For ",number," features: \n")
        print(transformed_data)
        print(transform_by_features.pvalues_)
        print(transform_by_features.scores_)
        
#    return transformed_data
possible_feature_numbers = [2,3,4,5,6,7,8,9,10]
featureSelectionOfKBestMeasure(standardized_generated_dataframe, generated_labels,possible_feature_numbers)


#Performing transformation to reduce dimensionality of the data
#transformed_generated_data = featureSelectionOfKBest(standardized_generated_dataframe, generated_labels, 9)
#print(transformed_generated_data)

SelectKBest(k=2)
For  2  features: 

[[ 1.39242057 -0.06182259]
 [-0.59950577  0.61543763]
 [-0.50615914  0.32379436]
 ...
 [ 0.7848968  -0.80211247]
 [-0.51528767 -0.92363826]
 [-0.00433482  0.88846665]]
[4.23954432e-002 0.00000000e+000 5.17241377e-001 6.65940780e-001
 8.69729652e-001 9.57887724e-001 0.00000000e+000 1.49275534e-001
 3.69397123e-001 3.47746340e-001 4.66626281e-001 7.78691123e-001
 7.68519842e-298 1.44738095e-001 7.17716362e-001 8.07975406e-003
 3.57050892e-001 4.15474137e-001 3.71510243e-001 6.71971623e-001
 9.21941157e-001 8.84510247e-001 1.16247328e-001 0.00000000e+000
 0.00000000e+000 0.00000000e+000 2.92792044e-001 3.35802420e-001
 2.57506869e-001 0.00000000e+000 6.55283704e-001 7.29505298e-251
 3.73092834e-001 1.74506608e-148 6.79508212e-001 7.23430662e-001
 7.25550925e-001 0.00000000e+000 3.84202661e-001 3.12104286e-001
 0.00000000e+000 3.23778881e-001 0.00000000e+000 3.21458487e-280
 2.39220867e-160 0.00000000e+000 7.46425140e-001 0.00000000e+000
 0.00000000e+00

Using the <code>SelectKBest</code> method with mutual info classifier metric to select the best eight features.
Mutual information is further described within the SKLearn documentation. [SKLearn docs on mutual information.](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.mutual_info_classif.html#sklearn.feature_selection.mutual_info_classif) The notes on mutual information that follow are paraphrased from SKLearn. 
Having a value of 0 is indicative of two random variables having no dependencies on one another. If the value is non-zero, that can be indicative of some dependencies between the two variables in question. The mutual information metric is based on KNN.

In [10]:
from sklearn.feature_selection import mutual_info_classif
def featureSelectionOfKBestResult(submitted_data, submittedLabels, focused_number):
    np.random.seed(90)
    transform_by_features = SelectKBest(score_func=mutual_info_classif,k=focused_number)
    transform_by_features.fit(X=submitted_data, y=submittedLabels)
    print(transform_by_features)
    transformed_data = transform_by_features.fit_transform(submitted_data, submittedLabels) 
    print(transformed_data)
    return transformed_data

transformed_data = featureSelectionOfKBestResult(standardized_generated_dataframe, generated_labels, 8)

SelectKBest(k=8, score_func=<function mutual_info_classif at 0x7a4b963a5cf0>)
[[ 1.42673235  1.4820108   0.67773855 ...  0.16507973  0.54125354
  -0.06182259]
 [-0.35328617 -0.33590322  0.44310246 ... -0.55629257 -0.83817178
   0.61543763]
 [-0.26103098 -0.73089161  0.05193602 ... -0.20552952  0.40152665
   0.32379436]
 ...
 [ 1.45604656  0.14106163 -0.02538978 ...  0.20212378  0.26068818
  -0.80211247]
 [-1.02946779  0.38289066  1.34609545 ...  0.52872378  1.04241002
  -0.92363826]
 [ 0.51273724 -0.58159775 -0.88431274 ... -0.6476322  -0.8382785
   0.88846665]]


<h2>Exploring the Data</h2>

Using the covariance matrix, we can observe if the whether the relevant features have a positive or negative relationship with one another.


In [11]:
# using ANOVA F-value to select the K best features of the data
from sklearn.covariance import empirical_covariance
print("Calculating the empirical covariance of the data:")
print(empirical_covariance(pd.DataFrame(data=generated_data)))
transformed_origin_generated_data = featureSelectionOfKBestResult(pd.DataFrame(data=generated_data), generated_labels, 8)
print("Calculating the empirical covariance of the transformed data:")
print(empirical_covariance(transformed_origin_generated_data))
print(transformed_origin_generated_data.shape)
 

Calculating the empirical covariance of the data:
[[ 9.94224004e-01 -2.75944244e-02  3.44190140e-03 ...  4.49019309e-03
   1.86278207e-02 -7.61682585e-03]
 [-2.75944244e-02  6.52016875e+00  3.71517903e-03 ... -1.39513595e-02
  -9.39814304e-01  1.30035465e-02]
 [ 3.44190140e-03  3.71517903e-03  9.94146514e-01 ... -1.22093149e-03
   9.02005266e-03 -5.80496093e-05]
 ...
 [ 4.49019309e-03 -1.39513595e-02 -1.22093149e-03 ...  1.00394354e+00
   4.72660595e-03 -3.57410271e-04]
 [ 1.86278207e-02 -9.39814304e-01  9.02005266e-03 ...  4.72660595e-03
   4.56502205e+00  4.07984927e-03]
 [-7.61682585e-03  1.30035465e-02 -5.80496093e-05 ... -3.57410271e-04
   4.07984927e-03  9.96977225e-01]]
SelectKBest(k=8, score_func=<function mutual_info_classif at 0x7a4b963a5cf0>)
[[ 2.89546808  3.3208769   1.60383671 ...  0.39309061  1.66679368
  -0.23464541]
 [-1.10837167 -0.27035596  1.18375352 ... -1.8917661  -2.02236295
   0.83898195]
 [-0.9008598  -1.0506433   0.48342452 ... -0.78076793  1.29310592
   0.376

<h2>Cross-Validation</h2>
<p>Implmenting the cross-validation for the selected model. This will be done for an assortment of models and their resulting performance will be accessed.</p>

In [12]:
def cross_validate_model_function(model,data,labels):
    resultsFromCV = cross_validate(estimator=model, X=data, y=labels,cv = 10,
                                   n_jobs = 3,
                                   #pre_dispatch =1,
                                   return_train_score=True, 
                                   scoring= {"accuracy": "accuracy", 
                                             "balanced_accuracy": "balanced_accuracy",
                                             "F1": "f1_micro"
                                            })
    return resultsFromCV

In [13]:
from sklearn.neural_network import MLPClassifier
mlp_neural_network_cl = MLPClassifier(hidden_layer_sizes=(8,4),activation="relu", solver="adam", learning_rate_init=0.01,max_iter=100,)
mlp_neural_network_cl_v2 = MLPClassifier(hidden_layer_sizes=(8,4),activation="relu", solver="adam", learning_rate_init=0.05,max_iter=100,)
from sklearn.model_selection import cross_validate


## MLP

<p>We will be using multilayer perceptron neural network below with <code>activation="relu"</code> and <code>solver=adam</code>. The neural network consists of an input layer that accepts the input data, the hidden layers where calculations are made to perform computations on supplied features, then the ouput layer that will produce the resulting prediction. Neural networks are a more complex predictive modeling method, and can be difficult to interpret.</p> 

In [12]:
mlp_neural_network_cross_validation = cross_validate_model_function(
    mlp_neural_network_cl,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
mlp_neural_network_cross_validation_v2 = cross_validate_model_function(
    mlp_neural_network_cl_v2,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)

In [13]:
import numpy as np
def get_model_scores(score_key_list, score_dict):
    for key in score_key_list:
        print(key," (median): " ,round(np.median(score_dict[key]),3))
        print(key, " (mean): ", round(np.mean(score_dict[key]),3), "\n")

def score_extract(score_dict):
    train_score_list = [key for key in score_dict if "train" in key]
    test_score_list = [key for key in score_dict if "test" in key]
    get_model_scores(train_score_list, score_dict)
    get_model_scores(test_score_list, score_dict)
    
print("The training and testing scores for first MLP, learning rate = 0.009: \n")
score_extract(mlp_neural_network_cross_validation)


print("\n The training and testing scores for second MLP, learning rate = 0.01: \n")
score_extract(mlp_neural_network_cross_validation_v2)

The training and testing scores for first MLP, learning rate = 0.009: 

train_accuracy  (median):  0.657
train_accuracy  (mean):  0.654 

train_balanced_accuracy  (median):  0.657
train_balanced_accuracy  (mean):  0.654 

train_F1  (median):  0.657
train_F1  (mean):  0.654 

test_accuracy  (median):  0.653
test_accuracy  (mean):  0.653 

test_balanced_accuracy  (median):  0.653
test_balanced_accuracy  (mean):  0.653 

test_F1  (median):  0.653
test_F1  (mean):  0.653 


 The training and testing scores for second MLP, learning rate = 0.01: 

train_accuracy  (median):  0.629
train_accuracy  (mean):  0.626 

train_balanced_accuracy  (median):  0.629
train_balanced_accuracy  (mean):  0.626 

train_F1  (median):  0.629
train_F1  (mean):  0.626 

test_accuracy  (median):  0.625
test_accuracy  (mean):  0.624 

test_balanced_accuracy  (median):  0.625
test_balanced_accuracy  (mean):  0.624 

test_F1  (median):  0.625
test_F1  (mean):  0.624 



<p>The KNN classifier is being used next. The models are being parallelized across three cores each.</p>

In [36]:
from sklearn.neighbors import KNeighborsClassifier
knn_model_cl_v1 = KNeighborsClassifier(n_neighbors =9, metric="euclidean",n_jobs = 3)
knn_model_cl_v2 = KNeighborsClassifier(n_neighbors =11, metric="euclidean",n_jobs = 3)
knn_model_cl_v3 = KNeighborsClassifier(n_neighbors =13, metric="euclidean",n_jobs = 3)
knn_model_cl_v4 = KNeighborsClassifier(n_neighbors =15, metric="euclidean",n_jobs = 3)
knn_model_cl_v1_validation =cross_validate_model_function(
    knn_model_cl_v1,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
knn_model_cl_v2_validation =cross_validate_model_function(
knn_model_cl_v2,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
knn_model_cl_v3_validation =cross_validate_model_function(
    knn_model_cl_v3,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
knn_model_cl_v4_validation =cross_validate_model_function(
    knn_model_cl_v4,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)

In [15]:
print("The result of KNN with 9 neighbors:")
print("The training and testing scores:")
score_extract(knn_model_cl_v1_validation)

print("The result of KNN with 11 neighbors:")
print("The training and testing scores:")
score_extract(knn_model_cl_v2_validation)

print("The result of KNN with 13 neighbors:")
print("The training and testing scores:")
score_extract(knn_model_cl_v3_validation)

print("The result of KNN with 15 neighbors:")
print("The training and testing scores:")
score_extract(knn_model_cl_v4_validation)

The result of KNN with 9 neighbors:
The training and testing scores:
train_accuracy  (median):  0.823
train_accuracy  (mean):  0.823 

train_balanced_accuracy  (median):  0.823
train_balanced_accuracy  (mean):  0.823 

train_F1  (median):  0.823
train_F1  (mean):  0.823 

test_accuracy  (median):  0.776
test_accuracy  (mean):  0.775 

test_balanced_accuracy  (median):  0.776
test_balanced_accuracy  (mean):  0.775 

test_F1  (median):  0.776
test_F1  (mean):  0.775 

The result of KNN with 11 neighbors:
The training and testing scores:
train_accuracy  (median):  0.815
train_accuracy  (mean):  0.816 

train_balanced_accuracy  (median):  0.815
train_balanced_accuracy  (mean):  0.816 

train_F1  (median):  0.815
train_F1  (mean):  0.816 

test_accuracy  (median):  0.778
test_accuracy  (mean):  0.776 

test_balanced_accuracy  (median):  0.778
test_balanced_accuracy  (mean):  0.776 

test_F1  (median):  0.778
test_F1  (mean):  0.776 

The result of KNN with 13 neighbors:
The training and tes

In [16]:
from sklearn.ensemble import GradientBoostingClassifier
gbc_ensemble_cl_v1 = GradientBoostingClassifier(n_estimators=5,learning_rate=0.215,random_state=3)
gbc_ensemble_cl_v2 = GradientBoostingClassifier(n_estimators=10,learning_rate=0.215,random_state=3)
gbc_ensemble_cl_v3 = GradientBoostingClassifier(n_estimators=15,learning_rate=0.215,random_state=3)
gbc_ensemble_cl_v4 = GradientBoostingClassifier(n_estimators=20,learning_rate=0.215,random_state=3)
gbc_ensemble_cl_v5 = GradientBoostingClassifier(n_estimators=50,learning_rate=0.215,random_state=3)
gbc_ensemble_cl_v1_validation = cross_validate_model_function(
    gbc_ensemble_cl_v1,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
gbc_ensemble_cl_v2_validation = cross_validate_model_function(
    gbc_ensemble_cl_v2,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
gbc_ensemble_cl_v3_validation = cross_validate_model_function(
    gbc_ensemble_cl_v3,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
gbc_ensemble_cl_v4_validation = cross_validate_model_function(
    gbc_ensemble_cl_v4,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
gbc_ensemble_cl_v5_validation = cross_validate_model_function(
    gbc_ensemble_cl_v5,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)


In [17]:
print("The result of Gradient Boost Classifier with 50 trees:")
print("The training and testing scores:")
score_extract(gbc_ensemble_cl_v1_validation)

print("The result of Gradient Boost Classifier with 100 trees:")
print("The training and testing scores:")
score_extract(gbc_ensemble_cl_v2_validation)

print("The result of Gradient Boost Classifier with 150 trees:")
print("The training and testing scores:")
score_extract(gbc_ensemble_cl_v3_validation)

print("The result of Gradient Boost Classifier with 200 trees:")
print("The training and testing scores:")
score_extract(gbc_ensemble_cl_v4_validation)

print("The result of Gradient Boost Classifier with 50 trees:")
print("The training and testing scores:")
score_extract(gbc_ensemble_cl_v5_validation)


The result of Gradient Boost Classifier with 50 trees:
The training and testing scores:
train_accuracy  (median):  0.558
train_accuracy  (mean):  0.558 

train_balanced_accuracy  (median):  0.558
train_balanced_accuracy  (mean):  0.558 

train_F1  (median):  0.558
train_F1  (mean):  0.558 

test_accuracy  (median):  0.552
test_accuracy  (mean):  0.552 

test_balanced_accuracy  (median):  0.552
test_balanced_accuracy  (mean):  0.552 

test_F1  (median):  0.552
test_F1  (mean):  0.552 

The result of Gradient Boost Classifier with 100 trees:
The training and testing scores:
train_accuracy  (median):  0.581
train_accuracy  (mean):  0.581 

train_balanced_accuracy  (median):  0.581
train_balanced_accuracy  (mean):  0.581 

train_F1  (median):  0.581
train_F1  (mean):  0.581 

test_accuracy  (median):  0.572
test_accuracy  (mean):  0.572 

test_balanced_accuracy  (median):  0.573
test_balanced_accuracy  (mean):  0.572 

test_F1  (median):  0.572
test_F1  (mean):  0.572 

The result of Gradi

In [18]:
from sklearn.linear_model import LogisticRegression
log_red_cl_v1 = LogisticRegression(penalty="l2",C=0.85, random_state = 3, solver="newton-cg", max_iter=200,n_jobs=2)
log_red_cl_v2 = LogisticRegression(penalty="l2",C=0.85, random_state = 3, solver="sag", max_iter=200,n_jobs=2)
log_red_cl_v3 = LogisticRegression(penalty="l2",C=0.75, random_state = 3, solver="newton-cg", max_iter=200,n_jobs=2)
log_red_cl_v4 = LogisticRegression(penalty="l2",C=0.75, random_state = 3, solver="sag", max_iter=200,n_jobs=2)
log_red_cl_v1_validation = cross_validate_model_function(
    log_red_cl_v1,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
log_red_cl_v2_validation = cross_validate_model_function(
    log_red_cl_v2,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
log_red_cl_v3_validation = cross_validate_model_function(
    log_red_cl_v3,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
log_red_cl_v4_validation = cross_validate_model_function(
    log_red_cl_v4,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)

In [19]:
def show_cv_train_test_results(results_dict):
    print("The training scores:")
    print(results_dict["train_score"])
    print("The testing scores:")
    print(results_dict["test_score"])

In [20]:
#Going into logistic regression
# Printing the training and testing scores
print("Logistic Regression train and test scores: ")
print("Newton solver logistic regression C = 0.85:" )
score_extract(log_red_cl_v1_validation)
print("SAG solverlogistic regression C = 0.85: ")
score_extract(log_red_cl_v2_validation)
print("Newton solver logistic regression C = 0.75:")
score_extract(log_red_cl_v3_validation)
print("SAG solverlogistic regression C = 0.75: ")
score_extract(log_red_cl_v4_validation)

Logistic Regression train and test scores: 
Newton solver logistic regression C = 0.85:
train_accuracy  (median):  0.477
train_accuracy  (mean):  0.477 

train_balanced_accuracy  (median):  0.477
train_balanced_accuracy  (mean):  0.477 

train_F1  (median):  0.477
train_F1  (mean):  0.477 

test_accuracy  (median):  0.476
test_accuracy  (mean):  0.476 

test_balanced_accuracy  (median):  0.476
test_balanced_accuracy  (mean):  0.476 

test_F1  (median):  0.476
test_F1  (mean):  0.476 

SAG solverlogistic regression C = 0.85: 
train_accuracy  (median):  0.477
train_accuracy  (mean):  0.477 

train_balanced_accuracy  (median):  0.477
train_balanced_accuracy  (mean):  0.477 

train_F1  (median):  0.477
train_F1  (mean):  0.477 

test_accuracy  (median):  0.476
test_accuracy  (mean):  0.476 

test_balanced_accuracy  (median):  0.476
test_balanced_accuracy  (mean):  0.476 

test_F1  (median):  0.476
test_F1  (mean):  0.476 

Newton solver logistic regression C = 0.75:
train_accuracy  (median

In [22]:
from sklearn.ensemble import BaggingClassifier
log_reg_cl_use_bagging_v1 = BaggingClassifier(estimator=log_red_cl_v1,n_estimators=5, n_jobs = 2 ,random_state=3)
log_reg_cl_use_bagging_v2 = BaggingClassifier(estimator=log_red_cl_v2,n_estimators=5, n_jobs = 2 ,random_state=3)

log_reg_cl_use_bagging_v1_validation = cross_validate_model_function(
    log_reg_cl_use_bagging_v1,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
log_reg_cl_use_bagging_v2_validation = cross_validate_model_function(
    log_reg_cl_use_bagging_v2,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)

print("Bagging with logistic regression Newton solver C= 0.85:")
print("Training and testing scores:")
score_extract(log_reg_cl_use_bagging_v1_validation)
print("Bagging with logistic regression SAG solver C= 0.85:")
print("Training and testing scores:")
score_extract(log_reg_cl_use_bagging_v2_validation)

Bagging with logistic regression Newton solver C= 0.85:
Training and testing scores:
train_accuracy  (median):  0.477
train_accuracy  (mean):  0.477 

train_balanced_accuracy  (median):  0.477
train_balanced_accuracy  (mean):  0.477 

train_F1  (median):  0.477
train_F1  (mean):  0.477 

test_accuracy  (median):  0.476
test_accuracy  (mean):  0.476 

test_balanced_accuracy  (median):  0.476
test_balanced_accuracy  (mean):  0.476 

test_F1  (median):  0.476
test_F1  (mean):  0.476 

Bagging with logistic regression SAG solver C= 0.85:
Training and testing scores:
train_accuracy  (median):  0.477
train_accuracy  (mean):  0.477 

train_balanced_accuracy  (median):  0.477
train_balanced_accuracy  (mean):  0.477 

train_F1  (median):  0.477
train_F1  (mean):  0.477 

test_accuracy  (median):  0.476
test_accuracy  (mean):  0.476 

test_balanced_accuracy  (median):  0.476
test_balanced_accuracy  (mean):  0.476 

test_F1  (median):  0.476
test_F1  (mean):  0.476 



In [23]:
from sklearn.ensemble import AdaBoostClassifier
log_reg_cl_adaboost_v1 = AdaBoostClassifier(estimator=log_red_cl_v1,n_estimators=15,random_state=3)
log_reg_cl_adaboost_v2 = AdaBoostClassifier(estimator=log_red_cl_v2,n_estimators=15,random_state=3)

log_reg_cl_adaboost_v1_validation = cross_validate_model_function(
    log_reg_cl_adaboost_v1,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)
log_reg_cl_adaboost_v2_validation = cross_validate_model_function(
    log_reg_cl_adaboost_v2,pd.DataFrame(data=transformed_origin_generated_data),generated_labels)

print("Adaboosting with Logistic Regression: ")
print("Bagging with logistic regression Newton solver C= 0.85:")
print("Training and testing scores:")
score_extract(log_reg_cl_adaboost_v1_validation)
print("Bagging with logistic regression SAG solver C= 0.85:")
print("Training and testing scores:")
score_extract(log_reg_cl_adaboost_v2_validation)

Adaboosting with Logistic Regression: 
Bagging with logistic regression Newton solver C= 0.85:
Training and testing scores:
train_accuracy  (median):  0.451
train_accuracy  (mean):  0.452 

train_balanced_accuracy  (median):  0.451
train_balanced_accuracy  (mean):  0.452 

train_F1  (median):  0.451
train_F1  (mean):  0.452 

test_accuracy  (median):  0.45
test_accuracy  (mean):  0.451 

test_balanced_accuracy  (median):  0.45
test_balanced_accuracy  (mean):  0.451 

test_F1  (median):  0.45
test_F1  (mean):  0.451 

Bagging with logistic regression SAG solver C= 0.85:
Training and testing scores:
train_accuracy  (median):  0.452
train_accuracy  (mean):  0.452 

train_balanced_accuracy  (median):  0.452
train_balanced_accuracy  (mean):  0.452 

train_F1  (median):  0.452
train_F1  (mean):  0.452 

test_accuracy  (median):  0.451
test_accuracy  (mean):  0.452 

test_balanced_accuracy  (median):  0.451
test_balanced_accuracy  (mean):  0.452 

test_F1  (median):  0.451
test_F1  (mean):  0

The KNN machine learning model was one of the top-performing. Note that Select K Best was done using 8 predictors, which introduced the possibility of potential predictors that were actually just noise being introduced within the machine learning models.
There are a lot of factors that could have influenced machine learning model performance. To better gauge each machine learning model, we can utilize the SKLearn pipeline utility along with grid search cross-validation to evaluate how well each machine learning model performs with a given set of hyperparameters.

## MLP Pipeline

In [14]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
# form a pipline for the MLP model
# start with the select K best with mutual information.
base_pipeline_mutual_info =[("mutual_info", SelectKBest(score_func=mutual_info_classif))]
mutual_info_params = {"mutual_info__k": [3, 5, 6, 7, 8, 10]}
# activation="relu", solver="adam", learning_rate_init=0.05,max_iter=100
#MLP_pipeline = base_pipeline_mutual_info.copy()
MLP_pipeline= base_pipeline_mutual_info + [("mlp_nn", MLPClassifier() )]
#MLP_pipeline = Pipeline(MLP_pipeline)
MLP_pipeline_params = mutual_info_params.copy()
MLP_pipeline_params.update({"mlp_nn__hidden_layer_sizes": [(8,3), (6,4), (10,2)],
                                                       "mlp_nn__activation": ["relu", "logistic"],
                                                       "mlp_nn__learning_rate_init": [0.01, 0.05, 0.1],
                                                       "mlp_nn__max_iter": [200,400]})

In [15]:
MLP_pipeline

[('mutual_info',
  SelectKBest(score_func=<function mutual_info_classif at 0x7a4b963a5cf0>)),
 ('mlp_nn', MLPClassifier())]

In [16]:
MLP_pipeline_params

{'mutual_info__k': [3, 5, 6, 7, 8, 10],
 'mlp_nn__hidden_layer_sizes': [(8, 3), (6, 4), (10, 2)],
 'mlp_nn__activation': ['relu', 'logistic'],
 'mlp_nn__learning_rate_init': [0.01, 0.05, 0.1],
 'mlp_nn__max_iter': [200, 400]}

In [23]:
MLP_grid_search = GridSearchCV(estimator = Pipeline(steps = MLP_pipeline), 
                               param_grid = MLP_pipeline_params, 
                               scoring= {"accuracy": "accuracy", 
                                             "balanced_accuracy": "balanced_accuracy",
                                             "F1": "f1_micro"
                                            }, refit = "F1",
                              n_jobs = 12,#5, 
                               cv = 10)

In [52]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(standardized_generated_dataframe, generated_labels, random_state = 34, test_size = 0.3, shuffle = True)

In [25]:
 y_train

array([2, 0, 0, ..., 3, 0, 1])

In [26]:
np.random.seed(12)
MLP_grid_search.fit(X = x_train, y = y_train)

/home/torvis/Documents/py_projects/ds_projects_scikit_tensorflow/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/torvis/Documents/py_projects/ds_projects_scikit_tensorflow/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/torvis/Documents/py_projects/ds_projects_scikit_tensorflow/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/torvis/Documents/py_projects/ds_projects_scikit_tensorflow/lib/python3.10/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: Con

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('mutual_info',
                                        SelectKBest(score_func=<function mutual_info_classif at 0x7a4b963a5cf0>)),
                                       ('mlp_nn', MLPClassifier())]),
             n_jobs=12,
             param_grid={'mlp_nn__activation': ['relu', 'logistic'],
                         'mlp_nn__hidden_layer_sizes': [(8, 3), (6, 4),
                                                        (10, 2)],
                         'mlp_nn__learning_rate_init': [0.01, 0.05, 0.1],
                         'mlp_nn__max_iter': [200, 400],
                         'mutual_info__k': [3, 5, 6, 7, 8, 10]},
             refit='F1',
             scoring={'F1': 'f1_micro', 'accuracy': 'accuracy',
                      'balanced_accuracy': 'balanced_accuracy'})

In [27]:
MLP_grid_search.best_params_

{'mlp_nn__activation': 'logistic',
 'mlp_nn__hidden_layer_sizes': (8, 3),
 'mlp_nn__learning_rate_init': 0.01,
 'mlp_nn__max_iter': 400,
 'mutual_info__k': 10}

In [28]:
# Best F1 Score
MLP_grid_search.best_score_

np.float64(0.6493928571428571)

In [29]:
logged_metrics = ["mean_test_accuracy", "std_test_accuracy", "mean_test_balanced_accuracy", "std_test_balanced_accuracy",
"mean_test_F1", "std_test_F1"]

In [30]:
for metric in logged_metrics:
    print("The overall average ", metric," is: ", round(np.mean(MLP_grid_search.cv_results_[metric]),3))
    print("The overall max ", metric," is: ", round(np.max(MLP_grid_search.cv_results_[metric]),3))
    print("The overall median ", metric," is: ", round(np.median(MLP_grid_search.cv_results_[metric]),3))
    print("-------------------------------------------------")

The overall average  mean_test_accuracy  is:  0.572
The overall max  mean_test_accuracy  is:  0.649
The overall median  mean_test_accuracy  is:  0.591
-------------------------------------------------
The overall average  std_test_accuracy  is:  0.026
The overall max  std_test_accuracy  is:  0.11
The overall median  std_test_accuracy  is:  0.018
-------------------------------------------------
The overall average  mean_test_balanced_accuracy  is:  0.572
The overall max  mean_test_balanced_accuracy  is:  0.649
The overall median  mean_test_balanced_accuracy  is:  0.59
-------------------------------------------------
The overall average  std_test_balanced_accuracy  is:  0.027
The overall max  std_test_balanced_accuracy  is:  0.11
The overall median  std_test_balanced_accuracy  is:  0.018
-------------------------------------------------
The overall average  mean_test_F1  is:  0.572
The overall max  mean_test_F1  is:  0.649
The overall median  mean_test_F1  is:  0.591
------------------

In [31]:
MLP_grid_search_predictions = MLP_grid_search.predict(X=x_test)

In [32]:
from sklearn.metrics import classification_report

In [55]:
print(classification_report(y_pred = MLP_grid_search_predictions, y_true = y_test))

              precision    recall  f1-score   support

           0       0.61      0.52      0.56      3041
           1       0.63      0.67      0.65      2994
           2       0.64      0.67      0.65      2962
           3       0.58      0.61      0.59      3003

    accuracy                           0.61     12000
   macro avg       0.61      0.62      0.61     12000
weighted avg       0.61      0.61      0.61     12000



We can observe that the MLP classifier suffered a bit. The imbalanced classes can be seen to have an effect on the MLP classifier as the 1st and 2nd classes have the best overall precision and one of the f1-scores. The 3rd class has the worst recall for all of the classes. In terms of the f1-score, the MLP classifier is more successful than guessing.

We can up the number of iterations to see if it resolves some issues with convergence.

In [45]:
# Second attempt
MLP_pipeline_params_att_2 = mutual_info_params.copy()
MLP_pipeline_params_att_2.update({"mlp_nn__hidden_layer_sizes": [(8,3), (6,4), (10,2)],
                                                       "mlp_nn__activation": ["relu", "logistic"],
                                                       "mlp_nn__learning_rate_init": [0.01, 0.05, 0.1],
                                                       "mlp_nn__max_iter": [800,1000]})

In [46]:
MLP_grid_search_att_2 = GridSearchCV(estimator = Pipeline(steps = MLP_pipeline), 
                               param_grid = MLP_pipeline_params_att_2, 
                               scoring= {"accuracy": "accuracy", 
                                             "balanced_accuracy": "balanced_accuracy",
                                             "F1": "f1_micro"
                                            }, refit = "F1",
                              n_jobs = 12,#5, 
                               cv = 10)

In [47]:
np.random.seed(12)
MLP_grid_search_att_2.fit(X = x_train, y = y_train)

GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('mutual_info',
                                        SelectKBest(score_func=<function mutual_info_classif at 0x7a4b963a5cf0>)),
                                       ('mlp_nn', MLPClassifier())]),
             n_jobs=12,
             param_grid={'mlp_nn__activation': ['relu', 'logistic'],
                         'mlp_nn__hidden_layer_sizes': [(8, 3), (6, 4),
                                                        (10, 2)],
                         'mlp_nn__learning_rate_init': [0.01, 0.05, 0.1],
                         'mlp_nn__max_iter': [800, 1000],
                         'mutual_info__k': [3, 5, 6, 7, 8, 10]},
             refit='F1',
             scoring={'F1': 'f1_micro', 'accuracy': 'accuracy',
                      'balanced_accuracy': 'balanced_accuracy'})

In [48]:
MLP_grid_search_att_2.best_params_

{'mlp_nn__activation': 'logistic',
 'mlp_nn__hidden_layer_sizes': (8, 3),
 'mlp_nn__learning_rate_init': 0.05,
 'mlp_nn__max_iter': 800,
 'mutual_info__k': 10}

In [49]:
for metric in logged_metrics:
    print("The overall average ", metric," is: ", round(np.mean(MLP_grid_search_att_2.cv_results_[metric]),3))
    print("The overall max ", metric," is: ", round(np.max(MLP_grid_search_att_2.cv_results_[metric]),3))
    print("The overall median ", metric," is: ", round(np.median(MLP_grid_search_att_2.cv_results_[metric]),3))
    print("-------------------------------------------------")

The overall average  mean_test_accuracy  is:  0.572
The overall max  mean_test_accuracy  is:  0.648
The overall median  mean_test_accuracy  is:  0.588
-------------------------------------------------
The overall average  std_test_accuracy  is:  0.028
The overall max  std_test_accuracy  is:  0.135
The overall median  std_test_accuracy  is:  0.019
-------------------------------------------------
The overall average  mean_test_balanced_accuracy  is:  0.571
The overall max  mean_test_balanced_accuracy  is:  0.648
The overall median  mean_test_balanced_accuracy  is:  0.588
-------------------------------------------------
The overall average  std_test_balanced_accuracy  is:  0.028
The overall max  std_test_balanced_accuracy  is:  0.135
The overall median  std_test_balanced_accuracy  is:  0.019
-------------------------------------------------
The overall average  mean_test_F1  is:  0.572
The overall max  mean_test_F1  is:  0.648
The overall median  mean_test_F1  is:  0.588
---------------

In [50]:
MLP_grid_search_att_2_predictions = MLP_grid_search_att_2.predict(X=x_test)

In [56]:
print(classification_report(y_pred = MLP_grid_search_att_2_predictions, y_true = y_test))

              precision    recall  f1-score   support

           0       0.62      0.68      0.65      3041
           1       0.62      0.68      0.65      2994
           2       0.66      0.68      0.67      2962
           3       0.71      0.54      0.61      3003

    accuracy                           0.65     12000
   macro avg       0.65      0.65      0.64     12000
weighted avg       0.65      0.65      0.64     12000



There were some improvements for classes 0 and 3 in terms of precision. Overall the f1-scores have improved. The neural network performed better with higher max iterations for convergence as expected.

## KNN Pipeline

In [37]:
# form a pipline for the KNN model
# start with the select K best with mutual information.
base_pipeline_mutual_info =[("mutual_info", SelectKBest(score_func=mutual_info_classif))]
mutual_info_params = {"mutual_info__k": [3, 5, 6, 7, 8, 10]}

KNN_pipeline= base_pipeline_mutual_info + [("knn", KNeighborsClassifier())]

KNN_pipeline_params = mutual_info_params.copy()
KNN_pipeline_params.update({"knn__n_neighbors": [3,5,7,11,13,15], # Avoid issues with ties
                                                       "knn__weights": ["uniform", "distance"]})

In [38]:
KNN_grid_search = GridSearchCV(estimator = Pipeline(steps = KNN_pipeline), 
                               param_grid = KNN_pipeline_params, 
                               scoring= {"accuracy": "accuracy", 
                                             "balanced_accuracy": "balanced_accuracy",
                                             "F1": "f1_micro"
                                            }, refit = "F1",
                              n_jobs = 12,#5, 
                               cv = 10)

In [39]:
np.random.seed(12)
KNN_grid_search.fit(X = x_train, y = y_train)

/home/torvis/Documents/py_projects/ds_projects_scikit_tensorflow/lib/python3.10/site-packages/numpy/ma/core.py:2881: RuntimeWarning: invalid value encountered in cast
  _data = np.array(data, dtype=dtype, copy=copy,


GridSearchCV(cv=10,
             estimator=Pipeline(steps=[('mutual_info',
                                        SelectKBest(score_func=<function mutual_info_classif at 0x7a4b963a5cf0>)),
                                       ('knn', KNeighborsClassifier())]),
             n_jobs=12,
             param_grid={'knn__n_neighbors': [3, 5, 7, 11, 13, 15],
                         'knn__weights': ['uniform', 'distance'],
                         'mutual_info__k': [3, 5, 6, 7, 8, 10]},
             refit='F1',
             scoring={'F1': 'f1_micro', 'accuracy': 'accuracy',
                      'balanced_accuracy': 'balanced_accuracy'})

In [40]:
# Get the best parameters overall
KNN_grid_search.best_params_

{'knn__n_neighbors': 11, 'knn__weights': 'distance', 'mutual_info__k': 10}

In [41]:
# Get the results from the Grid search cross-validation
for metric in logged_metrics:
    print("The overall average ", metric," is: ", round(np.mean(KNN_grid_search.cv_results_[metric]),3))
    print("The overall max ", metric," is: ", round(np.max(KNN_grid_search.cv_results_[metric]),3))
    print("The overall median ", metric," is: ", round(np.median(KNN_grid_search.cv_results_[metric]),3))
    print("-------------------------------------------------")

The overall average  mean_test_accuracy  is:  0.681
The overall max  mean_test_accuracy  is:  0.785
The overall median  mean_test_accuracy  is:  0.726
-------------------------------------------------
The overall average  std_test_accuracy  is:  0.016
The overall max  std_test_accuracy  is:  0.039
The overall median  std_test_accuracy  is:  0.014
-------------------------------------------------
The overall average  mean_test_balanced_accuracy  is:  0.681
The overall max  mean_test_balanced_accuracy  is:  0.785
The overall median  mean_test_balanced_accuracy  is:  0.726
-------------------------------------------------
The overall average  std_test_balanced_accuracy  is:  0.016
The overall max  std_test_balanced_accuracy  is:  0.039
The overall median  std_test_balanced_accuracy  is:  0.014
-------------------------------------------------
The overall average  mean_test_F1  is:  0.681
The overall max  mean_test_F1  is:  0.785
The overall median  mean_test_F1  is:  0.726
---------------

We get the results for the test set:

In [42]:
KNN_grid_search_predictions = KNN_grid_search.predict(X=x_test)

In [57]:
print(classification_report(y_pred = KNN_grid_search_predictions, y_true = y_test))

              precision    recall  f1-score   support

           0       0.76      0.85      0.80      3041
           1       0.81      0.79      0.80      2994
           2       0.80      0.76      0.78      2962
           3       0.82      0.79      0.80      3003

    accuracy                           0.80     12000
   macro avg       0.80      0.80      0.80     12000
weighted avg       0.80      0.80      0.80     12000



Overall, the results of the KNN classifier are far more accurate than that of the MLP classifier. It should be noted the MLP classifier had convergence issues with the number of max iterations set at 200. The second attempt with the MLP classifier using a larger numebr of iterations faired better overall. 

Generally, the f1-scores for classes 0, 1, and 3 are the highest for the KNN classifier. The precision for class 0 is 0.76. The recall for class 0 is 0.85. Class 2 may be suffering a bit in terms of the f1-score and recall as a result of being the minority class. We can see that overall, it is a good classifier.

Something to note is both the MLP and KNN under mutual information for feature selection are selecting ten features/predictors to use. Only six features are informative and 21 features are redundant within the simulated dataset.